File: app/src/main/java/com/ardabank/aradapay/data/remote/FirestoreService.kt
package com.ardabank.aradapay.data.remote

import com.ardabank.aradapay.domain.model.ApprovalStatus
import com.ardabank.aradapay.domain.model.CrossSettlementOffer
import com.ardabank.aradapay.domain.model.Expense
import com.ardabank.aradapay.domain.model.Nudge
import com.ardabank.aradapay.domain.model.Settlement
import com.ardabank.aradapay.domain.model.User
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FirestoreService @Inject constructor(
    private val db: FirebaseFirestore
) {
    // --- USER MANAGEMENT ---
    fun getUserFlow(userId: String): Flow<User?> = callbackFlow {
        val docRef = db.collection("users").document(userId)
        val listener = docRef.addSnapshotListener { snapshot, _ ->
            if (snapshot != null && snapshot.exists()) {
                trySend(snapshot.toObject(User::class.java))
            } else {
                trySend(null)
            }
        }
        awaitClose { listener.remove() }
    }

    fun getAllUsersFlow(): Flow<List<User>> = callbackFlow {
        val listener = db.collection("users").addSnapshotListener { snapshot, _ ->
            if (snapshot != null) {
                val users = snapshot.toObjects(User::class.java)
                trySend(users)
            } else {
                trySend(emptyList())
            }
        }
        awaitClose { listener.remove() }
    }

    fun saveUser(user: User, onComplete: (Boolean) -> Unit) {
        db.collection("users").document(user.id)
            .set(user)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    fun findUserByTag(tag: String, onResult: (User?) -> Unit) {
        db.collection("users")
            .whereEqualTo("tag", tag)
            .limit(1)
            .get()
            .addOnSuccessListener { snapshot ->
                if (snapshot != null && !snapshot.isEmpty) {
                    val user = snapshot.documents.firstOrNull()?.toObject(User::class.java)
                    onResult(user)
                } else {
                    onResult(null)
                }
            }
            .addOnFailureListener {
                onResult(null)
            }
    }

    fun syncFriendWithFirestore(currentUserId: String, friendUser: User, onComplete: (Boolean) -> Unit) {
        // Save friend record under user friends collection and save the friend user record in global directory
        db.collection("users").document(friendUser.id)
            .set(friendUser)
            .addOnSuccessListener {
                db.collection("users").document(currentUserId)
                    .collection("friends").document(friendUser.id)
                    .set(friendUser)
                    .addOnSuccessListener { onComplete(true) }
                    .addOnFailureListener { onComplete(false) }
            }
            .addOnFailureListener { onComplete(false) }
    }

    fun deleteUser(userId: String, onComplete: (Boolean) -> Unit) {
        db.collection("users").document(userId)
            .delete()
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    // --- EXPENSES ---
    fun getExpensesFlow(): Flow<List<Expense>> = callbackFlow {
        val listener = db.collection("expenses")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null) {
                    trySend(snapshot.toObjects(Expense::class.java))
                } else {
                    trySend(emptyList())
                }
            }
        awaitClose { listener.remove() }
    }

    fun addExpense(expense: Expense, onComplete: (Boolean) -> Unit) {
        db.collection("expenses").document(expense.id)
            .set(expense)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    fun updateExpenseStatus(expenseId: String, status: ApprovalStatus, onComplete: (Boolean) -> Unit) {
        db.collection("expenses").document(expenseId)
            .update("status", status)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    // --- SETTLEMENTS ---
    fun getSettlementsFlow(): Flow<List<Settlement>> = callbackFlow {
        val listener = db.collection("settlements")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null) {
                    trySend(snapshot.toObjects(Settlement::class.java))
                } else {
                    trySend(emptyList())
                }
            }
        awaitClose { listener.remove() }
    }

    fun addSettlement(settlement: Settlement, onComplete: (Boolean) -> Unit) {
        db.collection("settlements").document(settlement.id)
            .set(settlement)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    fun updateSettlementStatus(settlementId: String, status: ApprovalStatus, onComplete: (Boolean) -> Unit) {
        db.collection("settlements").document(settlementId)
            .update("status", status)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    // --- CROSS SETTLEMENT OFFERS ---
    fun getCrossSettlementOffersFlow(): Flow<List<CrossSettlementOffer>> = callbackFlow {
        val listener = db.collection("cross_settlement_offers")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null) {
                    trySend(snapshot.toObjects(CrossSettlementOffer::class.java))
                } else {
                    trySend(emptyList())
                }
            }
        awaitClose { listener.remove() }
    }

    fun createCrossSettlementOffer(offer: CrossSettlementOffer, onComplete: (Boolean) -> Unit) {
        db.collection("cross_settlement_offers").document(offer.id)
            .set(offer)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    fun updateCrossSettlementApproval(offerId: String, userId: String, approved: Boolean, onComplete: (Boolean) -> Unit) {
        db.collection("cross_settlement_offers").document(offerId)
            .update("approvals.$userId", approved)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    // --- NUDGES ---
    fun getNudgesFlow(userId: String): Flow<List<Nudge>> = callbackFlow {
        val listener = db.collection("nudges")
            .whereEqualTo("toUserId", userId)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null) {
                    trySend(snapshot.toObjects(Nudge::class.java))
                } else {
                    trySend(emptyList())
                }
            }
        awaitClose { listener.remove() }
    }

    fun sendNudge(nudge: Nudge, onComplete: (Boolean) -> Unit) {
        db.collection("nudges").document(nudge.id)
            .set(nudge)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    // --- GROUPS ---
    fun getGroupsFlow(): Flow<List<com.ardabank.aradapay.domain.model.Group>> = callbackFlow {
        val listener = db.collection("groups")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null) {
                    trySend(snapshot.toObjects(com.ardabank.aradapay.domain.model.Group::class.java))
                } else {
                    trySend(emptyList())
                }
            }
        awaitClose { listener.remove() }
    }

    fun saveGroup(group: com.ardabank.aradapay.domain.model.Group, onComplete: (Boolean) -> Unit) {
        db.collection("groups").document(group.id)
            .set(group)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }

    fun getGroupExpensesFlow(groupId: String): Flow<List<com.ardabank.aradapay.domain.model.GroupExpenseItem>> = callbackFlow {
        val listener = db.collection("groups").document(groupId)
            .collection("expenses")
            .orderBy("date", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                if (snapshot != null) {
                    trySend(snapshot.toObjects(com.ardabank.aradapay.domain.model.GroupExpenseItem::class.java))
                } else {
                    trySend(emptyList())
                }
            }
        awaitClose { listener.remove() }
    }

    fun addGroupExpense(groupId: String, expense: com.ardabank.aradapay.domain.model.GroupExpenseItem, onComplete: (Boolean) -> Unit) {
        db.collection("groups").document(groupId)
            .collection("expenses").document(expense.id)
            .set(expense)
            .addOnSuccessListener { onComplete(true) }
            .addOnFailureListener { onComplete(false) }
    }
}

File: web/src/services/firestoreService.ts
import {
  collection,
  doc,
  getDoc,
  getDocs,
  setDoc,
  updateDoc,
  deleteDoc,
  onSnapshot,
  query,
  orderBy,
  where,
  limit
} from 'firebase/firestore';
import { db } from '../firebase/config';
import {
  User,
  Expense,
  Settlement,
  Group,
  GroupExpenseItem,
  CrossSettlementOffer,
  Nudge,
  ApprovalStatus
} from '../types';

export const FirestoreService = {
  // --- USERS ---
  async getUser(userId: string): Promise<User | null> {
    try {
      const docRef = doc(db, 'users', userId);
      const snapshot = await getDoc(docRef);
      if (snapshot.exists()) {
        return snapshot.data() as User;
      }
      return null;
    } catch (e) {
      console.warn('Firestore getUser error:', e);
      return null;
    }
  },

  async findUserByTag(tag: string): Promise<User | null> {
    try {
      const q = query(collection(db, 'users'), where('tag', '==', tag), limit(1));
      const snapshot = await getDocs(q);
      if (!snapshot.empty) {
        return snapshot.docs[0].data() as User;
      }
      return null;
    } catch (e) {
      console.warn('Firestore findUserByTag error:', e);
      return null;
    }
  },

  subscribeUser(userId: string, callback: (user: User | null) => void) {
    try {
      const docRef = doc(db, 'users', userId);
      return onSnapshot(docRef, (snapshot) => {
        if (snapshot.exists()) {
          callback(snapshot.data() as User);
        } else {
          callback(null);
        }
      }, (err) => {
        console.warn('User subscription fallback to local', err);
      });
    } catch (e) {
      console.warn('Firestore subscribeUser error:', e);
      return () => {};
    }
  },

  subscribeAllUsers(callback: (users: User[]) => void) {
    try {
      const colRef = collection(db, 'users');
      return onSnapshot(colRef, (snapshot) => {
        const users = snapshot.docs.map((d) => d.data() as User);
        callback(users);
      }, (err) => {
        console.warn('Users collection listener fallback', err);
      });
    } catch (e) {
      console.warn('Firestore subscribeAllUsers error:', e);
      return () => {};
    }
  },

  async saveUser(user: User): Promise<boolean> {
    try {
      await setDoc(doc(db, 'users', user.id), user);
      return true;
    } catch (e) {
      console.error('saveUser error:', e);
      return false;
    }
  },

  async deleteUser(userId: string): Promise<boolean> {
    try {
      await deleteDoc(doc(db, 'users', userId));
      return true;
    } catch (e) {
      console.error('deleteUser error:', e);
      return false;
    }
  },

  // --- EXPENSES ---
  subscribeExpenses(callback: (expenses: Expense[]) => void) {
    try {
      const q = query(collection(db, 'expenses'), orderBy('createdAt', 'desc'));
      return onSnapshot(q, (snapshot) => {
        const expenses = snapshot.docs.map((d) => d.data() as Expense);
        callback(expenses);
      }, (err) => {
        console.warn('Expenses listener fallback', err);
      });
    } catch (e) {
      console.warn('Firestore subscribeExpenses error:', e);
      return () => {};
    }
  },

  async addExpense(expense: Expense): Promise<boolean> {
    try {
      await setDoc(doc(db, 'expenses', expense.id), expense);
      return true;
    } catch (e) {
      console.error('addExpense error:', e);
      return false;
    }
  },

  async updateExpenseStatus(expenseId: string, status: ApprovalStatus): Promise<boolean> {
    try {
      await updateDoc(doc(db, 'expenses', expenseId), { status });
      return true;
    } catch (e) {
      console.error('updateExpenseStatus error:', e);
      return false;
    }
  },

  // --- SETTLEMENTS ---
  subscribeSettlements(callback: (settlements: Settlement[]) => void) {
    try {
      const q = query(collection(db, 'settlements'), orderBy('createdAt', 'desc'));
      return onSnapshot(q, (snapshot) => {
        const settlements = snapshot.docs.map((d) => d.data() as Settlement);
        callback(settlements);
      }, (err) => {
        console.warn('Settlements listener fallback', err);
      });
    } catch (e) {
      console.warn('Firestore subscribeSettlements error:', e);
      return () => {};
    }
  },

  async addSettlement(settlement: Settlement): Promise<boolean> {
    try {
      await setDoc(doc(db, 'settlements', settlement.id), settlement);
      return true;
    } catch (e) {
      console.error('addSettlement error:', e);
      return false;
    }
  },

  // --- GROUPS ---
  subscribeGroups(callback: (groups: Group[]) => void) {
    try {
      const q = query(collection(db, 'groups'), orderBy('createdAt', 'desc'));
      return onSnapshot(q, (snapshot) => {
        const groups = snapshot.docs.map((d) => d.data() as Group);
        callback(groups);
      }, (err) => {
        console.warn('Groups listener fallback', err);
      });
    } catch (e) {
      console.warn('Firestore subscribeGroups error:', e);
      return () => {};
    }
  },

  async saveGroup(group: Group): Promise<boolean> {
    try {
      await setDoc(doc(db, 'groups', group.id), group);
      return true;
    } catch (e) {
      console.error('saveGroup error:', e);
      return false;
    }
  },

  subscribeGroupExpenses(groupId: string, callback: (expenses: GroupExpenseItem[]) => void) {
    try {
      const q = query(collection(db, 'groups', groupId, 'expenses'), orderBy('date', 'desc'));
      return onSnapshot(q, (snapshot) => {
        const expenses = snapshot.docs.map((d) => d.data() as GroupExpenseItem);
        callback(expenses);
      }, (err) => {
        console.warn('Group expenses fallback', err);
      });
    } catch (e) {
      console.warn('subscribeGroupExpenses error:', e);
      return () => {};
    }
  },

  async addGroupExpense(groupId: string, expense: GroupExpenseItem): Promise<boolean> {
    try {
      await setDoc(doc(db, 'groups', groupId, 'expenses', expense.id), expense);
      return true;
    } catch (e) {
      console.error('addGroupExpense error:', e);
      return false;
    }
  },

  // --- CROSS SETTLEMENT OFFERS ---
  subscribeCrossOffers(callback: (offers: CrossSettlementOffer[]) => void) {
    try {
      const q = query(collection(db, 'cross_settlement_offers'), orderBy('createdAt', 'desc'));
      return onSnapshot(q, (snapshot) => {
        const offers = snapshot.docs.map((d) => d.data() as CrossSettlementOffer);
        callback(offers);
      }, (err) => {
        console.warn('Cross offers fallback', err);
      });
    } catch (e) {
      console.warn('Firestore subscribeCrossOffers error:', e);
      return () => {};
    }
  },

  async createCrossOffer(offer: CrossSettlementOffer): Promise<boolean> {
    try {
      await setDoc(doc(db, 'cross_settlement_offers', offer.id), offer);
      return true;
    } catch (e) {
      console.error('createCrossOffer error:', e);
      return false;
    }
  },

  async updateCrossApproval(offerId: string, userId: string, approved: boolean): Promise<boolean> {
    try {
      await updateDoc(doc(db, 'cross_settlement_offers', offerId), {
        [`approvals.${userId}`]: approved
      });
      return true;
    } catch (e) {
      console.error('updateCrossApproval error:', e);
      return false;
    }
  },

  // --- NUDGES ---
  subscribeNudges(userId: string, callback: (nudges: Nudge[]) => void) {
    try {
      const q = query(collection(db, 'nudges'), where('toUserId', '==', userId));
      return onSnapshot(q, (snapshot) => {
        const nudges = snapshot.docs.map((d) => d.data() as Nudge);
        callback(nudges);
      }, (err) => {
        console.warn('Nudges listener fallback', err);
      });
    } catch (e) {
      console.warn('Firestore subscribeNudges error:', e);
      return () => {};
    }
  },

  async sendNudge(nudge: Nudge): Promise<boolean> {
    try {
      await setDoc(doc(db, 'nudges', nudge.id), nudge);
      return true;
    } catch (e) {
      console.error('sendNudge error:', e);
      return false;
    }
  }
};
